### ISSUE TITLE
ACC Lap 1 S1 issue when time is inaccurate

### ISSUE THREAD

**Issue by FxUK** [mauserrifle/simresults#76]
Hi Mauserrifle, 

I noticed an issue with ACC results, whereby the first sector of lap one can sometimes be rather large, as in 30 minutes.
Because this is included, it then proceeds to throws off a lot of stats, such as gaps by lap, positions by lap, etc.

As an example, in the result file that produced this, it appears drivers S1 time for l1 of a race includes when they joined in practice, so some have 30 mins, some 10 mins etc.  i.e
https://cdn.discordapp.com/attachments/682933545211396140/807984352003620904/unknown.png

Gaps by lap is therefore, way off the mark.
https://cdn.discordapp.com/attachments/682933545211396140/807993851314176021/unknown.png

I made a quick hack whereby if S1 is > 2 mins, treat it as 60s just experimenting with potential fixes (which this likely isn't ideal) and it did indeed correct all the relevant stats, such as positions by lap/gaps by lap etc.
https://cdn.discordapp.com/attachments/682933545211396140/807993918959910932/unknown.png

However, I'm not sure if its an elegant solution or not, so thought I would give you a little nudge and let you know about the issue.

**Comment by mauserrifle** [comment]
Hi FxUK,

Thanks for reporting! It's indeed a known problem. One of the ACC developers told me:

> I don't think we can do something about it at this stage, the timer starts when the player enters the session or presses drive, I think the final results normalize this by removing the first sector if I remember correctly.

So I think I will do the same for Simresults. Funny enough I never fixed it and kind of forgot about it! Your > 2 min isn't safe due long laps on some tracks (like Nordschleife etc). A safer solution would be some sort of percentage calculation based on average lap time. But that won't cut it either.

I am going to look at it again today. I think I will just remove s1 of lap 1 :)

Thanks for reminding me :smiley:

**Comment by mauserrifle** [comment]
Ok it got quite complicated. The total time in the logs also includes the sector 1 time. So I do not know the real total time and it is also bugged (very high times).

I did not like excluding lap 1 fully. So I have created the following:

All lap 1  sector 1 times are now based on the sector 1 average of laps 2+ including a increment of 5 seconds to simulate the grid start additional time (must be done if all averages are based on fast alien drivers). This correction only happens for race sessions.

Please test! :smiley: 

1b634ecbb0b8759d8628bb5e9cdf0f9b909ad6bc

**Comment by FxUK** [comment]
:) it's not an easy one, for sure.  My quick hack was just a first thought, but yes, its not safe especially when long tracks are in use.  
I will give it a try shortly, once again.. thanks for everything.. a great tool so far!

**Comment by FxUK** [comment]
I think you have found a reasonable solution - I have implemented the changes and all looks good, to me. 👍

**Comment by FxUK** [comment]
Hi Mauserrifle, 

So, it transpires that this isn't actually working as well as I first thought, there are still oddities depending on how that first sector goes (i.e incidents / slow starts etc).  However, after banging our heads together (in our club)... I believe we have a good idea that may resolve the S1 issue.

If you take the driver total time, then deduct all laps (or sectors) excluding S1, this should give you the correct S1 time.
What are your thoughts.

**Comment by FxUK** [comment]
The DNF's kind of make this not a very good solution either, although close than the first.
I'm wondering if working it out backwards, since we know the finish position.. whether that would have the desired result.
The issue is positions per lap are all wrong when a dnf occurs

**Comment by mauserrifle** [comment]
Ugh this is giving me a headache :rofl:  It is so hard to dig in again. I am going to ask the developers of ACC again to really consider adding the proper time. Because I suspect this will keep us busy.

Since it is for me hard to dig in again with the current explanation. If you are convinced you found a good solution. Please explain me with real data:

1. a log file
2. a single driver picked as use case
3. how it is calculated
4. example code (if available)

This way it is easier for me to dig in again.

For now I hope I can convince the developers to just fix it or as a quickfix add a new s1 field.

**Comment by mauserrifle** [comment]
Can't calculate with grand totalTime because it may contain penalties. It is to be expected it is not always the sum of all laps.

**Comment by FxUK** [comment]
Ok, I think I have something... although again, not in code, but in theory 

Remove S1 and get the averages as you have.
Then instead of adding 5 seconds to each, add in increments of 0.0001 to each first lap.
Each lap is output in the json in order of crossing the line, so as a really rough example

$incr = 0.001 ;
$applied = array() ;

foreach( $laps as $lap ) { 

//Get driver ID for lap
$driverID = . . . . . ;

if( ! in_array($driverID, $appliedAdjustments)) {
laps[s1Time] = $average + $incr ;
$incr = $incr + 0.001 ;
$applied[] = $driverID ;
}

if( all participants accounted for) {
break;
}
}

If my theory is correct, this should ensure all drivers cross the line for Lap 1 in the correct order.. and whilst the gaps per lap themselves may be slightly off, they should still be relative, and the positions by lap should be accurate.

**Comment by mauserrifle** [comment]
Thanks for thinking this out! I made the following changes: 

https://github.com/mauserrifle/simresults/commit/b7a490064a93c09e466f7ff76efa93bd2753cd57
1. Set the position for lap 1 only. Because in the results I tested the lap order seemed bugged. I had jumps that didn't reflect the gap times (driver in front of other regarding position, but behind in gaps). So setting the first position is just for fixing s1.
2. Added the `0.001` increments to the first sector

This seems to improve my graphs a bit. But I am sure you got some proper tests.

Please test! :) Again thanks for your brainwork!

PS. I also tried using `leaderBoardLines` json order for the grid start position. But it seems in finish order. So that didn't work out.

**Comment by FxUK** [comment]
I implemented the same (or v. similar at least) and there's a definite improvement, your code is of course cleaner :) so will replace my file a bit later.  I didn't want you to think I was hassling, I just kept having brain waves lol.
 
Anyway, It corrected quite a few issues that we had with a few result files and whilst there's still some scope for it to have a knock on effect, this is minimised a lot compared to previously.

Thanks for also testing out the lap order, I never got around to trying it so far, I couldn't figure out how to deal with lapping cars / dnfs etc.

The only true fix would have to come from ACC, but its definitely much much better now.
